Sources/swift-package-editor/SwiftPackageEditorTool.swift

```diff
@@ -0,0 +1,345 @@
+/*
+ This source file is part of the Swift.org open source project
+
+ Copyright (c) 2021 Apple Inc. and the Swift project authors
+ Licensed under Apache License v2.0 with Runtime Library Exception
+
+ See http://swift.org/LICENSE.txt for license information
+ See http://swift.org/CONTRIBUTORS.txt for Swift project authors
+*/
+
+import ArgumentParser
+import Basics
+import TSCBasic
+import PackageModel
+import PackageGraph
+import SourceControl
+import Workspace
+import Foundation
+import PackageSyntax
+import PackageLoading
+import TSCUtility
+
+@main
+public struct SwiftPackageEditorTool: ParsableCommand {
+    public static var configuration = CommandConfiguration(
+        commandName: "package-editor",
+        _superCommandName: "swift",
+        abstract: "Edit Package.swift files",
+        version: SwiftVersion.current.completeDisplayString,
+        subcommands: [AddDependency.self, AddTarget.self, AddProduct.self],
+        helpNames: [.short, .long, .customLong("help", withSingleDash: true)])
+    public init() {}
+}
+
+final class EditorTool {
+    let diagnostics: DiagnosticsEngine
+    let packageRoot: AbsolutePath
+    let toolchain: UserToolchain
+    let packageEditor: PackageEditor
+    private var cachedPackageGraph: PackageGraph?
+
+    init() throws {
+        diagnostics = DiagnosticsEngine(handlers: [print(diagnostic:)])
+
+        guard let cwd = localFileSystem.currentWorkingDirectory else {
+            diagnostics.emit(.error("could not determine current working directory"))
+            throw ExitCode.failure
+        }
+
+        var root = cwd
+        while !localFileSystem.isFile(root.appending(component: Manifest.filename)) {
+            root = root.parentDirectory
+            guard !root.isRoot else {
+                diagnostics.emit(.error("could not find package manifest"))
+                throw ExitCode.failure
+            }
+        }
+        packageRoot = root
+
+        toolchain = try UserToolchain(destination: Destination.hostDestination(originalWorkingDirectory: cwd))
+
+        let manifestPath = try ManifestLoader.findManifest(packagePath: packageRoot,
+                                                           fileSystem: localFileSystem,
+                                                           currentToolsVersion: ToolsVersion.current)
+        let repositoryManager = RepositoryManager(fileSystem: localFileSystem,
+                                                  path: packageRoot.appending(component: ".build"),
+                                                  provider: GitRepositoryProvider(),
+                                                  initializationWarningHandler: { _ in })
```

Comment:
feels like this can be refactored to lean more on code from other tools / SwiftTool